crypto/bls12381_utils.c

```diff
@@ -99,7 +99,7 @@ ctx_t* relic_init_BLS12_381() {
     if (ALLOC != AUTO) return NULL;
 
     // initialize relic core with a new context
-    ctx_t* bls_ctx = (ctx_t*) malloc(sizeof(ctx_t));
+    ctx_t* bls_ctx = (ctx_t*) calloc(1, sizeof(ctx_t));
```

Comment:
@tarakby while here, also changed `malloc` to `calloc` since the latter zero-initializes the context.  ([given that default context](https://github.com/relic-toolkit/relic/blob/main/src/relic_core.c#L99) is `static` I would assume it expects that)  This fixes `msan` complaint:
```
==2776351==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x14a7aef in bls_verify_ep /home/rbtz/tmp/go/pkg/mod/github.com/onflow/flow-go/crypto@v0.24.3/bls_core.c:145:9
    #1 0x14a7898 in bls_verify /home/rbtz/tmp/go/pkg/mod/github.com/onflow/flow-go/crypto@v0.24.3/bls_core.c:328:12
    #2 0x14a1184 in _cgo_01218c0baead_Cfunc_bls_verify /tmp/go-build/cgo-gcc-prolog:75:11
    #3 0x527bef in runtime.asmcgocall.abi0 /home/rbtz/tmp/go/src/runtime/asm_amd64.s:765

SUMMARY: MemorySanitizer: use-of-uninitialized-value /home/rbtz/tmp/go/pkg/mod/github.com/onflow/flow-go/crypto@v0.24.3/bls_core.c:145:9 in bls_verify_ep
Exiting
```

Reply:
Interesting catch!

May I know how you  run the sanitizer?

From a quick look at the code, I think the warning is a sign of another racing issue in the code. I suspect it is not caught because the race Go tooling doesn't analyze things beyond Cgo.  The context  is a shared and not protected struct, initialized by the `init()` of the Go package and used by the function in logs `bls_verify_ep`. Is it possible that a thread is accessing the [freshly created context](https://github.com/SaveTheRbtz/flow-go/blob/combined/crypto/bls12381_utils.c#L103) before it gets [initialized](https://github.com/SaveTheRbtz/flow-go/blob/combined/crypto/bls12381_utils.c#L104) ?

I can dig further into that, and look generally at racing issues in the crypto tests, though I'll create another issue for that work.

Reply:
I was using the following test:
```
$ CC=/usr/bin/clang CXX=/usr/bin/clang++ go test -c ./engine/execution/ingestion -tags=relic -msan
$ ./ingestion.test
==2776351==WARNING: MemorySanitizer: use-of-uninitialized-value
...
```
From what I can understand, it is complaining about: ```core_get()->code == RLC_OK```.  That said, I can't find the path where `code` can be uninitialized on `bls_verify`, unless it is called between these two lines:
```
    core_set(bls_ctx);
    if (core_init() != RLC_OK) return NULL;
```

Reply:
The problem is that IIUC `lsInstance.init()` is called in the `init` function of the module, so it should not be able to race.... unless, maybe, there are some background threads created by the relic itself?  Or maybe one of flow's `init` functions launches a goroutine?  Though in these cases I would assume that bug should not be reliably triggered...

Reply:
> unless it is called between these two lines

Yes exactly, that's what I think. The fix you added (setting to zero) makes the second line check valid, which I think is a happy coincidence, but the racing issue might still be there.   

>is called in the init function of the module, so it should not be able to race

That was my understanding too, but not after your test :) (I do not think Relic is running anything in the backgroud)

I cannot run `-msan` because it's only supported on `linux`, so we can keep `calloc` for now, and I'll find a way to test later. 
Thank you again for catching this 🙏🏼

Reply:
Seems like there is also a potential race condition there: 
```
$ CC=/usr/bin/clang CXX=/usr/bin/clang++ CGO_CFLAGS="-fsanitize=thread" CGO_LDFLAGS="-fsanitize=thread" ~/go/bin/dlv test --build-flags="-tags=relic" ./engine/execution/ingestion
...
WARNING: ThreadSanitizer: data race (pid=3757976)
  Read of size 8 at 0x000005738ae0 by thread T9:
    #0 fp_get_sign /home/rbtz/go/pkg/mod/github.com/onflow/flow-go/crypto@v0.24.3/bls12381_utils.c:239:26 (debug.test+0x15ac0a3)
    #1 ep_write_bin_compact /home/rbtz/go/pkg/mod/github.com/onflow/flow-go/crypto@v0.24.3/bls12381_utils.c:272:24 (debug.test+0x15abf9c)
    #2 bls_sign_ep /home/rbtz/go/pkg/mod/github.com/onflow/flow-go/crypto@v0.24.3/bls_core.c:82:5 (debug.test+0x15adda8)
    #3 bls_sign /home/rbtz/go/pkg/mod/github.com/onflow/flow-go/crypto@v0.24.3/bls_core.c:93:5 (debug.test+0x15adcdb)
    #4 _cgo_01218c0baead_Cfunc_bls_sign /tmp/go-build/cgo-tsan-prolog:33:2 (debug.test+0x15a9aec)
    #5 runtime.asmcgocall.abi0 /usr/lib/go/src/runtime/asm_amd64.s:821 (debug.test+0x5676a3)

  Previous write of size 8 at 0x000005738ae0 by thread T8:
    #0 precomputed_data_set /home/rbtz/go/pkg/mod/github.com/onflow/flow-go/crypto@v0.24.3/bls12381_utils.c:54:14 (debug.test+0x15ab550)
    #1 _cgo_01218c0baead_Cfunc_precomputed_data_set /tmp/go-build/cgo-tsan-prolog:336:2 (debug.test+0x15aa855)
    #2 runtime.asmcgocall.abi0 /usr/lib/go/src/runtime/asm_amd64.s:821 (debug.test+0x5676a3)

  Location is global 'bls_prec' of size 8 at 0x000005738ae0 (debug.test+0x000005738ae0)```

Likely because `blsInstance.reInit()` is called unconditionally on each sign/verify.  Theoretically it should be safe, since this data should static after `init()` but if it is -- is there a reason to call it there?  If it is not static, then it can potentially race between different signers/verifiers.

Reply:
Agree, that is another race condition that I'm aware of for some time : there is a single static relic context, initialized with a BLS-friendly context during `init()`. The crypto library today uses only one setup (context) but could be evolved to use more contexts in the future. Whenever a BLS related function is called, no assumptions about the Relic context value are made, so `reinit()` is called to set up the BLS context each time. This is obviously not thread safe, but since the lib for now is based only on one context, the same context value is being assigned over and over, and the race condition doesn't cause errors. 

A temporary solution would be to not call `reinit()` at all because the call is obsolete (I can do this in another PR). If more contexts are supported in the future, then a thread safe solution needs to be implemented.

Reply:
I created this [issue](https://github.com/dapperlabs/flow-internal/issues/1699)   to address these two race conditions. You'll get access to the repo soon :) Thanks for reporting these points.